### ISSUE TITLE
Different image behavior in Google Container Builder vs locally

### ISSUE THREAD

**Issue by svachmic** [GoogleCloudPlatform/cloud-builders#142]
We have a Spring Boot application (multiple micro-services), packaged with Google Container Builder and published to the registry. Our `cloudbuild.yaml` looks like this:

```
steps:
- name: 'gcr.io/cloud-builders/java/gradle'
  id: 'java-build'
  args: ['build']

- name: 'gcr.io/cloud-builders/docker'
  args: ['build', 
         '-t', 'gcr.io/$PROJECT_ID/api-auth:$COMMIT_SHA',
         '-t', 'gcr.io/$PROJECT_ID/api-auth:latest',
         '-f', './api-auth/Dockerfile', './api-auth']
  waitFor: ['java-build']
- name: 'gcr.io/cloud-builders/docker'
  args: ['build', 
         '-t', 'gcr.io/$PROJECT_ID/api-user:$COMMIT_SHA',
         '-t', 'gcr.io/$PROJECT_ID/api-user:latest',
         '-f', './api-user/Dockerfile', './api-user']
  waitFor: ['java-build']
 
images:
- 'gcr.io/$PROJECT_ID/api-auth:$COMMIT_SHA'
- 'gcr.io/$PROJECT_ID/api-auth:latest'
- 'gcr.io/$PROJECT_ID/api-user:$COMMIT_SHA'
- 'gcr.io/$PROJECT_ID/api-user:latest'
```

which is called by a trigger on push into the repository on GitHub. 

If I run this locally:

```
container-builder-local --dryrun=false .
```

resulting images run fine, e.g.:

```
docker run -e "SPRING_PROFILES_ACTIVE=development" -p 9000:9000 gcr.io/.../api-auth:latest
```

but when the image is built in the cloud with the Container Builder, the application doesn't work and fails with error that never ocurred before.

Even when pulled:

```
gcloud docker -- pull gcr.io/.../api-auth:latest
```

and run locally the new error stops the application from loading:

```
2017-10-11 16:35:27.661  WARN 5 --- [           main] ationConfigEmbeddedWebApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'webSecurityConfig': Unsatisfied dependency expressed through field 'authenticationManager'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'authenticationManager' defined in class path resource [org/springframework/boot/autoconfigure/security/AuthenticationManagerConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.authentication.AuthenticationManager]: Factory method 'authenticationManager' threw exception; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'OAuth2Config': Unsatisfied dependency expressed through field 'authenticationManager'; nested exception is org.springframework.beans.factory.BeanCurrentlyInCreationException: Error creating bean with name 'authenticationManager': Requested bean is currently in creation: Is there an unresolvable circular reference?
2017-10-11 16:35:27.663  INFO 5 --- [           main] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2017-10-11 16:35:27.676  INFO 5 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
2017-10-11 16:35:27.718  INFO 5 --- [           main] utoConfigurationReportLoggingInitializer :

Error starting ApplicationContext. To display the auto-configuration report re-run your application with 'debug' enabled.
2017-10-11 16:35:27.730 ERROR 5 --- [           main] o.s.b.d.LoggingFailureAnalysisReporter   :

***************************
APPLICATION FAILED TO START
***************************

Description:

The dependencies of some of the beans in the application context form a cycle:

   webSecurityConfig (field private org.springframework.security.authentication.AuthenticationManager com.project.auth.config.WebSecurityConfig.authenticationManager)
┌─────┐
|  authenticationManager defined in class path resource [org/springframework/boot/autoconfigure/security/AuthenticationManagerConfiguration.class]
↑     ↓
|  OAuth2Config (field private org.springframework.security.authentication.AuthenticationManager com.project.auth.config.OAuth2Config.authenticationManager)
└─────┘
```

This only happens when the image goes through the Cloud Container Builder, not the local one, not with local gradle build.

Could this possibly be a problem in the cloud, or am I missing something trivial?

**Comment by Philmod** [comment]
Hey @svachmic ,

That is the first time we hear this issue. Let's try to find out what's happening.

1. When you run the image locally, do you use the one created in your `docker images`? Or you download it from GCR? [0]
2. Do you use a `.dockerignore` file? Maybe there are some dependencies from your local machine that shouldn't go to the GCB worker? [1]

[0] With the local builder, if you don't pass the flag `--push=true`, the images won't be pushed.
[1] I'm not familiar with java, but with nodejs, if you have your dependencies installed on a local mac (`node_modules` directory) and you send it along with your code to GCB, it will fail because it's not the same os.

**Comment by svachmic** [comment]
Hi @Philmod thanks for reaching out!

1. I tested both - local and pulled from the GCR. Local always runs, GCR always fails.
2. No `.dockerignore`.

I haven't thought about the dependencies, but surely, if the jar is built in a fresh gradle image, the resulting image should be the same regardless of the platform. Both the local and the cloud builders use the same images for building our artifacts. The only thing I don't see is the underlying container "packaging" system. My guess is that Google doesn't use Docker, but something in-house with Docker compatibility, right?

**Comment by Philmod** [comment]
1. Can you delete your local image (`docker rmi ...`) and retry to build with local builder? Just to make sure the image you are testing doesn't come from somewhere else. Actually, I would try to delete most docker images you have locally first.
2. Not sure if `.dockerignore` makes sense for java, just throwing ideas.

We do use docker on GCB, we actually use the same code as the local builder (with some added magic). This is why I'm very surprised with this problem.

**Comment by meltsufin** [comment]
Can you trying diffing the local and the remote images with https://github.com/GoogleCloudPlatform/container-diff?

**Comment by svachmic** [comment]
Hi everyone,

sorry for the long period of silence. The problem disappeared by itself and never came back. I was unable to reproduce it. Right now Something similar has occurred - it works locally, local docker image works well, but once it goes through CloudBuilder it stops working and crashes on strange error that just makes no sense.

I ran the images through the container-diff and this is the result I got. I hope it brings any value:

**latest** is the image pulled from cloud container registry, **local** is built locally

```
❯ container-diff diff daemon://gcr.io/project/api-auth:latest daemon://gcr.io/project/api-auth:local --type=history --type=file
Starting diff on images daemon://gcr.io/project/api-auth:latest and daemon://gcr.io/project/api-auth:local, using differs: [history file]
Retrieving image gcr.io/project/api-auth:local from source Local Daemon
Retrieving image gcr.io/project/api-auth:latest from source Local Daemon
Computing diffsERRO[0027] Error checking directory entry /lib64/ld-linux-x86-64.so.2: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/lib64/ld-linux-x86-64.so.2: no such file or directory

ERRO[0027] Error checking directory entry /usr/bin/java: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/bin/java: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/debug/usr/lib/jvm/java-1.8.0-openjdk-amd64: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/debug/usr/lib/jvm/java-1.8.0-openjdk-amd64: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/accessibility.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/accessibility.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/jvm.cfg: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/jvm.cfg: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/calendars.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/calendars.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/content-types.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/content-types.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/flavormap.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/flavormap.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/images/cursors/cursors.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/images/cursors/cursors.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/logging.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/logging.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/jmxremote.access: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/jmxremote.access: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/jmxremote.password: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/jmxremote.password: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/management.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/management.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/snmp.acl: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management/snmp.acl: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/net.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/net.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/psfont.properties.ja: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/psfont.properties.ja: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/psfontj2d.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/psfontj2d.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/blacklisted.certs: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/blacklisted.certs: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/java.policy: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/java.policy: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/java.security: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/java.security: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/nss.cfg: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/nss.cfg: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/sound.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/sound.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/swing.properties: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/swing.properties: no such file or directory

ERRO[0027] Error checking directory entry /usr/lib/ssl/private: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/lib/ssl/private: no such file or directory

ERRO[0028] Error checking directory entry /usr/share/doc/libgcc1: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/share/doc/libgcc1: no such file or directory

ERRO[0028] Error checking directory entry /usr/share/doc/libgomp1: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/share/doc/libgomp1: no such file or directory

ERRO[0028] Error checking directory entry /usr/share/doc/libstdc++6: stat /var/folders/xc/gy3c62216r73tlh7c_ykhc1c0000gn/T/gcr.ioprojectapi-authlatest256208093/usr/share/doc/libstdc++6: no such file or directory


-----File-----

These entries have been added to gcr.io/project/api-auth:latest: None

These entries have been deleted from gcr.io/project/api-auth:latest: None

These entries have been changed between gcr.io/project/api-auth:latest and gcr.io/project/api-auth:local:
FILE            SIZE1        SIZE2
/app.jar        56.3M        56.3M


-----History-----

Docker history lines found only in gcr.io/project/api-auth:latest:
-/bin/sh -c #(nop) ADD file:1191ac9a81d37f90279271af19a4c226716c6fc8b8dc10004557f6e6ea15cd84 in app.jar


Docker history lines found only in gcr.io/project/api-auth:local:
-/bin/sh -c #(nop) ADD file:e3105582fd0dd0a46b2934b2ae329716886d90ee277e26053809f32e2844b7f2 in app.jar
```

**Comment by Philmod** [comment]
So you are saying that when you build your server image with local builder or simply with Docker, your server works; but it doesn't work if you build the image with GCB?

**Comment by svachmic** [comment]
That's exactly right. I know it sounds strange and believe me - I've triple checked the last possible different parts (the injected cloudbuild env variables, but I didn't touch those for weeks). But still this is the behavior I have.

**Comment by Philmod** [comment]
1. What Docker version are you using locally?
2. Does the GCB build succeed or fail?
3. Could you give me a `buildId` for one of those builds?

**Comment by ImJasonH** [comment]
We've been talking this through a bit on the Slack channel. There's still plenty more to debug, but we've ruled at least some things out:

* Using Docker 17.12 on macOS locally, he's going to try to reproduce using 17.06 on a Cloud Shell instance
* Build `787d0ff3-6d42-4910-9201-81ce81c207f8` is the last one that works on GCB, `eb00db2d-02c6-471d-9b7d-8c30354750b0` doesn't.
* He's tried pinning to the exact step digests from the working build, and it still fails.

**Comment by sebgoa** [comment]
I have a similar issue.

the local `container-builder-local --dryrun=false .` runs fine without errors.

But the remote build job fails.

**Comment by sebgoa** [comment]
my previous comment in this thread was fixed in #293
